### PR TITLE
[#1177] improvement: Reduce the write time of tasks

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?
Use callback and queue to optimize the write time.

### Why are the changes needed?

Fix: #1177

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
CI passed.

### CODE REVIEW

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 1]
```diff
@@ -293,24 +297,52 @@ protected List<CompletableFuture<Long>> postBlockEvent(
...
+          remainingMs = end - System.currentTimeMillis();
+          if (remainingMs < 0) {
+            remainingMs = 0;
+          }
+        }
       }
-      LOG.info("Wait " + blockIds.size() + " blocks sent to shuffle server");
-      Uninterruptibles.sleepUninterruptibly(sendCheckInterval, TimeUnit.MILLISECONDS);
-      if (System.currentTimeMillis() - start > sendCheckTimeout) {
+      if (remainingMs <= 0) {
```
Comment (xianjingfeng) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:345]: How about `end - System.currentTimeMillis() <=0 `? I think we can remove L339 to L342, right?

Comment (jerqi) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:345]: OK.

**Review by xianjingfeng — DISMISSED** [review]
LGTM

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 2]
```diff
@@ -293,24 +297,47 @@ protected List<CompletableFuture<Long>> postBlockEvent(
...
   @VisibleForTesting
   protected void checkBlockSendResult(Set<Long> blockIds) {
-    long start = System.currentTimeMillis();
-    while (true) {
-      checkIfBlocksFailed();
-      Set<Long> successBlockIds = shuffleManager.getSuccessBlockIds(taskId);
-      blockIds.removeAll(successBlockIds);
-      if (blockIds.isEmpty()) {
-        break;
+    boolean interrupted = false;
```
Comment (zuston) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:314]: It looks complex for current implement.

And I think the sendTimeout is wrong, which is valid for all everytime `queue.poll`

Comment (jerqi) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:314]: We use `remainingMs` instead of `sendCheckTImeout`.

Comment (zuston) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:314]: My fault. +1

**Review by zuston — APPROVED** [review]

**Pull request merged**